clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2206,18 +2251,20 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
     return true;
   }
 
-  bool nextElement(QualType ET) final {
-    CurOffset += SemaRef.getASTContext().getTypeSizeInChars(ET).getQuantity();
+  bool enterArray() final {
+    ArrayBases.push_back(CurOffset);
     return true;
   }
 
-  bool leaveArray(FieldDecl *, QualType ET, int64_t Count) final {
-    int64_t ArraySize =
-        SemaRef.getASTContext().getTypeSizeInChars(ET).getQuantity();
-    if (!ET->isArrayType()) {
-      ArraySize *= Count;
-    }
-    CurOffset -= ArraySize;
+  bool nextElement(QualType ET, int64_t Index) final {
+    int64_t Size = SemaRef.getASTContext().getTypeSizeInChars(ET).getQuantity();
+    CurOffset = ArrayBases.back() + Size * (Index + 1);
```

Comment:
Is this +1 right? I would think that the offset of element 0 is still at 0, right?

Reply:
I don't think element 0 enters this. It is handled in enterArray. I will take another look and confirm.

Reply:
But then Index 1 would be at 2*Size, when I would expect it to be at 1* size, right?

Reply:
Index is basically passed from VisitArrayElements right after we process an element. E.g. After processing the first element, we call nextElement to get the offset for it but its still passing 0 as index (the index for element we just finished processing) which I agree is confusing. 
```
for (int64_t Count = 0; Count < ElemCount; Count++) {
      VisitElement(nullptr, FD, ET, handlers...);
      (void)std::initializer_list<int>{(handlers.nextElement(ET, Count), 0)...};
 }
```
I'll update VisitArrayElements to pass Count+1 instead.

Reply:
Thats probably good enough for now.  In reality, 'nextElement' should probably be called before VisitElement...

Reply:
^^ That is likely a bigger change than we want to deal with here by the way.  I think having it be Count +1 for now is probably the right answer for now.